### PR TITLE
fix: add support for with-expecter when using generics

### PR DESCRIPTION
**Pull request by LouisBrunner**
Description
-------------

When generating generic mocks with `with-expecter: true`, the generated code is invalid and does not build. This means codebase which have generics cannot use `with-expecter: true` (which is a shame!).

This PR fixes this issue without any breaking changes.

One consideration: I had to add 2 extra calls to `getInstantiatedTypeString` and `getTypeConstraintString`, not sure if this could have some impact on performance and should be performed once for the whole file or if 3 times per file is ok.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

Version of Golang used when building/testing:
---------------------------------------------

- [ ] 1.11
- [ ] 1.12
- [ ] 1.13
- [ ] 1.14
- [ ] 1.15
- [ ] 1.16
- [ ] 1.17
- [x] 1.18

How Has This Been Tested?
---------------------------

Added an extra unit test in `generator_test.go`, tested on my company codebase which includes 2 different generic types and loads of other interfaces.

Checklist
-----------

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- ~[ ] I have commented my code, particularly in hard-to-understand areas~ none required (?)
- ~[ ] I have made corresponding changes to the documentation~ none required
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by LandonTClipp** [comment]
Thanks for the PR! I'll review this in the coming days.

File: pkg/generator_test.go [hunk 1]
```diff
@@ -2346,6 +2346,190 @@ func NewRequesterGenerics[TAny interface{}, TComparable comparable, TSigned cons
...
+type RequesterGenerics_GenericAnonymousStructs_Call[TAny interface{}, TComparable comparable, TSigned constraints.Signed, TIntf test.GetInt, TExternalIntf io.Writer, TGenIntf test.GetGeneric[TSigned], TInlineType interface{ ~int | ~uint }, TInlineTypeGeneric interface {
+	~int | test.GenericType[int, test.GetInt]
+	comparable
+}] struct {
+	*mock.Call
+}
+
+// GenericAnonymousStructs is a helper method to define mock.On call
+//  - _a0 struct{Type1 TExternalIntf}
+func (_e *RequesterGenerics_Expecter[TAny, TComparable, TSigned, TIntf, TExternalIntf, TGenIntf, TInlineType, TInlineTypeGeneric]) GenericAnonymousStructs(_a0 interface{}) *RequesterGenerics_GenericAnonymousStructs_Call[TAny, TComparable, TSigned, TIntf, TExternalIntf, TGenIntf, TInlineType, TInlineTypeGeneric] {
```
Comment (LandonTClipp) [inline pkg/generator_test.go:2401]:
```
GenericAnonymousStructs(_a0 interface{})
```

Can you explain this bit? Should `_a0` have a type of `struct{Type1 TExternalIntf}`?

**Comment by LandonTClipp** [comment]
@LouisBrunner I think this looks good. If you could, can you provide an example of this code producing the expecter interface with your changes, and you being able to use the mocks? Just as a last sanity check. I wish the tests had a way to do this programmatically (I don't really like the "just check that the generated strings match" thing) but alas.

**Comment by LouisBrunner** [comment]
Sure, here is a tiny example I put together: [mockery.zip](https://github.com/vektra/mockery/files/9018936/mockery.zip)

**Comment by LandonTClipp** [comment]
That's great, tests/example look good to me. Thanks!

**Pull request merged**